big_tests/tests/vcard_SUITE.erl

```diff
@@ -1123,20 +1142,37 @@ restart_vcard_mod(Config, ldap_only) ->
 restart_vcard_mod(Config, _GN) ->
     restart_mod(params_all(Config)).
 
-start_running_vcard_mod(Config) ->
-    Domain = ct:get_config({hosts, mim, domain}),
-    OriginalVcardConfig = ?config(mod_vcard, Config),
-    dynamic_modules:start(Domain, mod_vcard, OriginalVcardConfig).
-stop_running_vcard_mod(Config) ->
-    Domain = ct:get_config({hosts, mim, domain}),
-    CurrentConfigs = rpc(mim(), gen_mod, loaded_modules_with_opts, [Domain]),
-    {mod_vcard, CurrentVcardConfig} = lists:keyfind(mod_vcard, 1, CurrentConfigs),
-    dynamic_modules:stop(Domain, mod_vcard),
-    [{mod_vcard, CurrentVcardConfig} | Config].
+restart_mod(Params) ->
+    [restart_mod(HostType, Params) || HostType <- host_types()],
+    ok.
+
+restart_mod(HostType, Params) ->
+    dynamic_modules:stop(HostType, mod_vcard),
+    {ok, _Pid} = dynamic_modules:start(HostType, mod_vcard, Params).
+
+prepare_vcard_module(Config) ->
+    %% Keep the old config, so we can undo our changes, once finished testing
+    Config1 = dynamic_modules:save_modules_for_host_types(host_types(), Config),
+    %% Get a list of options, we can use as a prototype to start new modules
+    [{mod_vcard, get_vcard_config(Config)} | Config1].
+
+restore_vcard_module(Config) ->
+    dynamic_modules:restore_modules(Config).
+
+host_types() ->
+    HostType = ct:get_config({hosts, mim, host_type}),
+    SecHostType = ct:get_config({hosts, mim, secondary_host_type}),
+    lists:usort([HostType, SecHostType]).
+
+get_vcard_config(Config) ->
+    CfgHost = ct:get_config({hosts, mim, configured_host}),
```

Comment:
I don't get this, why not something like `domain_helper:host_type()`?

Reply:
Because mod_vcard is not configured for "test type".

```
gen_mod:hosts_and_opts_with_module(mod_vcard).
[{<<"anonymous.localhost">>,
  [{backend,rdbms},{host,{prefix,<<"vjud.">>}}]},
 {<<"localhost">>,
  [{backend,rdbms},{host,{prefix,<<"vjud.">>}}]},
 {<<"localhost.bis">>,
  [{backend,rdbms},{host,{prefix,<<"vjud.">>}}]}]
 ```
  
```
./tools/test-runner.sh --preset pgsql_mnesia --skip-cover --skip-small-tests --skip-stop-nodes --spec dynamic_domains.spec vcard
```

```erlang
=== Ended at 2021-08-20 11:53:38
=== Location: [{distributed_helper,rpc,117},
              {vcard_SUITE,prepare_vcard_module,1157},
              {vcard_SUITE,init_per_suite,127},
              {test_server,ts_tc,1754},
              {test_server,run_test_case_eval1,1351},
              {test_server,run_test_case_eval,1195}]
=== === Reason: {badrpc,
                     {'EXIT',
                         {{module_not_loaded,<<"test type">>,mod_vcard},
                          [{gen_mod,get_loaded_module_opts,2,
                               [{file,
                                    "/Users/mikhailuvarov/erlang/esl/MongooseIM/src/gen_mod.erl"},
                                {line,375}]},
                           {rpc,'-handle_call_call/6-fun-0-',5,
                               [{file,"rpc.erl"},{line,197}]}]}}}
  in function  distributed_helper:rpc/4
     called as distributed_helper:rpc(#{node => mongooseim@localhost},
                                      gen_mod,get_loaded_module_opts,
                                      [<<"test type">>,mod_vcard])
  in call from vcard_SUITE:prepare_vcard_module/1 (vcard_SUITE.erl, line 1157)
  in call from vcard_SUITE:init_per_suite/1 (vcard_SUITE.erl, line 127)
  in call from test_server:ts_tc/3 (test_server.erl, line 1754)
  in call from test_server:run_test_case_eval1/6 (test_server.erl, line 1351)
  in call from test_server:run_test_case_eval/9 (test_server.erl, line 1195)
=== 
*** init_per_suite failed.
    Skipping all cases.
```

Reply:
I think it's better to configure it for "test type" just like other modules (e.g. `mod_roster`, `mod_last`, `mod_offline`, `mod_privacy`) than to introduce this `configured_host` (type). I even think that the `dynamic_domains` preset should work correctly without `localhost` configured at all.